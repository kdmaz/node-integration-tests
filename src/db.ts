import { Pool } from 'pg'

const connectionString =
  'postgresql://postgres:postgres@localhost:5433/integration'

export const pool = new Pool({ connectionString })

export function closeDb(): void {
  pool.end()
}
