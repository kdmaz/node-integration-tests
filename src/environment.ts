import dotenv from 'dotenv'

const nodeEnv = process.env.NODE_ENV
const path = nodeEnv ? `.env.${process.env.NODE_ENV}` : '.env'

dotenv.config({ path })

export const env = {
  connectionString: process.env.DB_CONNECTION_STRING,
}
