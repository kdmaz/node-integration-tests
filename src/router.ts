import express from 'express'
import { router as todoController } from './todos/todo.controller'
export const router = express.Router()

router.use('/todos', todoController)
