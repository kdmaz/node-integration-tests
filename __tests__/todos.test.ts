import supertest from 'supertest'
import { app } from '../src/app'
import { pool } from '../src/db'
import { TodoDto } from '../src/todos/todo.interface'

const request = supertest(app)

const baseUrl = '/api/todos'

const todos: TodoDto[] = [
  {
    desc: 'mow the grass',
    id: 1,
    isComplete: false,
  },
  {
    desc: 'take out the garbage',
    id: 2,
    isComplete: true,
  },
]

async function insertTodos(): Promise<void> {
  for (let i = 0; i < todos.length; i++) {
    const { desc, isComplete } = todos[i]
    await pool.query({
      text: 'INSERT INTO todo (description, is_complete) VALUES ($1, $2)',
      values: [desc, isComplete],
    })
  }
}

async function clearTodos(): Promise<void> {
  await pool.query('DROP TABLE IF EXISTS todo')
  await pool.query(`CREATE TABLE IF NOT EXISTS todo (
    todo_id SERIAL PRIMARY KEY,
    description VARCHAR(50) NOT NULL,
    is_complete BOOLEAN NOT NULL
  )`)
}
beforeEach(async () => {
  await clearTodos()
  await insertTodos()
})

afterAll(() => {
  pool.end()
})

test('get all todos', async () => {
  const { body: todos, statusCode } = await request.get(baseUrl)
  expect(statusCode).toBe(200)
  expect(todos).toMatchSnapshot()
})

test('add a todo', async () => {
  const {
    body: { todoId },
    statusCode: postStatusCode,
  } = await request.post(baseUrl).send({
    desc: 'new todo',
    isComplete: true,
  })
  expect(postStatusCode).toBe(201)
  expect(todoId).toBe(3)

  const { body: todos, statusCode: getStatusCode } = await request.get(baseUrl)
  expect(getStatusCode).toBe(200)
  expect(todos).toMatchSnapshot()
})

test('delete a todo', async () => {
  const { body, statusCode: deleteStatusCode } = await request.delete(
    `${baseUrl}/1`,
  )
  expect(deleteStatusCode).toBe(204)
  expect(body).toEqual({})

  const { body: todos, statusCode: getStatusCode } = await request.get(baseUrl)
  expect(getStatusCode).toBe(200)
  expect(todos).toMatchSnapshot()
})
