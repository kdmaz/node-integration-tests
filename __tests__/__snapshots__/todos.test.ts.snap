// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add a todo 1`] = `
Array [
  Object {
    "desc": "mow the grass",
    "id": 1,
    "isComplete": false,
  },
  Object {
    "desc": "take out the garbage",
    "id": 2,
    "isComplete": true,
  },
  Object {
    "desc": "new todo",
    "id": 3,
    "isComplete": true,
  },
]
`;

exports[`delete a todo 1`] = `
Array [
  Object {
    "desc": "take out the garbage",
    "id": 2,
    "isComplete": true,
  },
]
`;

exports[`get all todos 1`] = `
Array [
  Object {
    "desc": "mow the grass",
    "id": 1,
    "isComplete": false,
  },
  Object {
    "desc": "take out the garbage",
    "id": 2,
    "isComplete": true,
  },
]
`;
